src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -5073,6 +5073,110 @@ .maxstack  1
+
+                void assertExpected(string param1Type, string param2Type, string returnType, string fieldName)
+                {
+                    var field = (FunctionPointerTypeSymbol)c.GetField(fieldName).Type;
+
+                    var paramTypes = field.Signature.ParameterTypesWithAnnotations;
+                    Assert.Equal(2, paramTypes.Length);
+                    Assert.Equal(param1Type, paramTypes[0].ToTestDisplayString(includeNonNullable: true));
+                    Assert.Equal(param2Type, paramTypes[1].ToTestDisplayString(includeNonNullable: true));
+
+                    Assert.Equal(returnType, field.Signature.ReturnTypeWithAnnotations.ToTestDisplayString(includeNonNullable: true));
+                }
+            }
+        }
+
+        [Theory]
+        [InlineData("delegate*<Z, Z, (Z a, Z b)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({""a"", ""b""})")]
+        [InlineData("delegate*<(Z a, Z b), Z, Z>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({""a"", ""b""})")]
+        [InlineData("delegate*<Z, (Z a, Z b)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({""a"", ""b""})")]
+        [InlineData("delegate*<(Z c, Z d), (Z e, Z f), (Z a, Z b)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({""a"", ""b"", ""c"", ""d"", ""e"", ""f""})")]
+        [InlineData("delegate*<(Z, Z), (Z, Z), (Z a, Z b)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({""a"", ""b"", null, null, null, null})")]
+        [InlineData("delegate*<(Z, Z), (Z, Z), (Z a, Z)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({""a"", null, null, null, null, null})")]
+        [InlineData("delegate*<(Z, Z), (Z, Z), (Z, Z b)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({null, ""b"", null, null, null, null})")]
+        [InlineData("delegate*<(Z c, Z d), (Z, Z), (Z, Z)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({null, null, ""c"", ""d"", null, null})")]
+        [InlineData("delegate*<(Z c, Z), (Z, Z), (Z, Z)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({null, null, ""c"", null, null, null})")]
+        [InlineData("delegate*<(Z, Z d), (Z, Z), (Z, Z)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({null, null, null, ""d"", null, null})")]
+        [InlineData("delegate*<(Z, Z), (Z e, Z f), (Z, Z)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({null, null, null, null, ""e"", ""f""})")]
+        [InlineData("delegate*<(Z, Z), (Z e, Z), (Z, Z)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({null, null, null, null, ""e"", null})")]
+        [InlineData("delegate*<(Z, Z), (Z, Z f), (Z, Z)>", @"System.Runtime.CompilerServices.TupleElementNamesAttribute({null, null, null, null, null, ""f""})")]
+        [InlineData("delegate*<(Z, Z), (Z, Z), (Z, Z)>", null)]
+        public void TupleNamesInMetadata(string type, string? expectedAttribute)
```

Comment:
>public [](start = 8, length = 6)

Consider testing some nested cases:
```
delegate*<(Z a, (Z b, Z c) d), (Z e, Z f)>
delegate*<(Z a, Z b), ((Z c, Z d) e, Z f)>
delegate*<(Z a, delegate*<(Z b, Z c)> d), delegate*<(Z e, Z f)>>
```

Reply:
That last one isn't valid code, as function pointers can't be type arguments. But I can do the others, and I can add one where the nested function pointers have tuples.

---
In reply to: [420570493](https://github.com/dotnet/roslyn/pull/43961#discussion_r420570493) [](ancestors = 420570493)